File: nodejs/commons/src/gcp/scheduler.ts
import { CloudSchedulerClient } from '@google-cloud/scheduler';
import { LoggerFactory } from '@day1co/pebbles';

const logger = LoggerFactory.getLogger('cornerstone-starter-commons-starter');

export interface SetupSchedulerOptions {
  projectId: string;
  region: string;
  jobName: string;
  topicName: string;
  // endpoint: string;
  schedule: string;
  timeZone: string;
}

export async function setupScheduler({
  projectId,
  region,
  jobName,
  topicName,
  schedule,
  timeZone,
}: SetupSchedulerOptions) {
  const client = new CloudSchedulerClient();

  const request = {
    parent: client.locationPath(projectId, region),
    job: {
      name: jobName,
      // $ gcloud scheduler jobs create pubsub myjob
      // --schedule "0 1 * * 0"
      // --topic cron-topic
      // --message-body "Hello"
      pubsubTarget: { topicName },
      // $ gcloud scheduler jobs create http test-job --schedule "5 * * * *"
      // --http-method=HTTP-METHOD
      // --uri=SERVICE-URL
      //httpTarget: {
      //  uri: endpoint,
      //  httpMethod: 'POST',
      //  body: JSON.stringify({ message: Buffer.from(message).toString('base64') }),
      //},
      schedule,
      timeZone,
    },
  };
  logger.debug('creating scheduler job: %o', request);

  const [job] = await client.createJob(request);
  logger.debug('scheduler job created: %o', job);

  return job;
}

File: java/commons/src/main/java/com/day1co/deimos/commons/gcp/Scheduler.java
package com.day1co.cornerstone.commons.gcp;

import com.google.cloud.scheduler.v1.*;
import com.google.protobuf.ByteString;
import lombok.Builder;
import lombok.Data;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.function.Consumer;

public class Scheduler implements Consumer<Scheduler.Command> {
    private static final Logger logger = LoggerFactory.getLogger(Scheduler.class);

    @Data
    @Builder
    public static class Command {
        public String project;
        public String region;
        public String job;
        public String endpoint;
        public String schedule;
        public String timeZone;
    }

    /**
     * 정해진 일정에 따라 토픽에 메시지를 전송할 cloud scheduler job 생성
     */
    @Override
    public void accept(Command command) {
        try (var cloudSchedulerClient = CloudSchedulerClient.create()) {
            var parent = LocationName.of(command.project, command.region).toString();
            // $ gcloud scheduler jobs create pubsub myjob
            // --schedule "0 1 * * 0"
            // --topic cron-topic
            // --message-body "Hello"
            // var pubsubTarget = PubsubTarget.newBuilder()
            //        .setTopicName(command.topic)
            //        .setData(ByteString.EMPTY)
            //        .build();
            // $ gcloud scheduler jobs create http test-job --schedule "5 * * * *"
            // --http-method=HTTP-METHOD
            // --uri=SERVICE-URL
            // --body
            var httpTarget = HttpTarget.newBuilder()
                    .setUri(command.endpoint)
                    .setHttpMethod(HttpMethod.POST)
                    .setBody(ByteString.EMPTY) // JSON.stringify({ message: Buffer.from(message).toString('base64') }),
                    .build();
            var job = Job.newBuilder()
                    .setName(command.job)
                    //.setPubsubTarget(pubsubTarget)
                    .setHttpTarget(httpTarget)
                    .setSchedule(command.schedule)
                    .setTimeZone(command.timeZone)
                    .build();
            var request = CreateJobRequest.newBuilder()
                    .setParent(parent)
                    .setJob(job)
                    .build();
            var result = cloudSchedulerClient.createJob(request);
            logger.info("scheduler job created: {}", result);
        } catch (Throwable t) {
            logger.warn("**ignore** failed to create scheduler job:" + command.job, t);
        }
    }
}
